src/ripple/app/tx/impl/SetRemarks.cpp

```diff
@@ -0,0 +1,451 @@
+#include <ripple/protocol/PublicKey.h>
+#include <ripple/protocol/Quality.h>
+#include <ripple/protocol/st.h>
+
+namespace ripple {
+
+TxConsequences
+SetRemarks::makeTxConsequences(PreflightContext const& ctx)
+{
+    return TxConsequences{ctx.tx, TxConsequences::normal};
+}
+
+NotTEC
+SetRemarks::validateRemarks(STArray const& remarks, beast::Journal const& j)
+{
+    std::set<Blob> already_seen;
+
+    if (remarks.empty() || remarks.size() > 32)
+    {
+        JLOG(j.warn()) << "SetRemarks: Cannot set more than 32 remarks (or "
+                          "fewer than 1) in a txn.";
+        return temMALFORMED;
+    }
+
+    for (auto const& remark : remarks)
+    {
+        if (remark.getFName() != sfRemark)
+        {
+            JLOG(j.warn()) << "SetRemarks: contained non-sfRemark field.";
+            return temMALFORMED;
+        }
+
+        // will be checked by template system, extra check for security
+        if (!remark.isFieldPresent(sfRemarkName))
+            return temMALFORMED;
+
+        Blob const& name = remark.getFieldVL(sfRemarkName);
+        if (already_seen.find(name) != already_seen.end())
+        {
+            JLOG(j.warn()) << "SetRemarks: duplicate RemarkName entry.";
+            return temMALFORMED;
+        }
+        if (name.size() == 0 || name.size() > 256)
+        {
+            JLOG(j.warn()) << "SetRemarks: RemarkName cannot be empty or "
+                              "larger than 256 chars.";
+            return temMALFORMED;
+        }
+
+        already_seen.emplace(name);
+
+        uint32_t flags =
+            remark.isFieldPresent(sfFlags) ? remark.getFieldU32(sfFlags) : 0;
+        if (flags != 0 && flags != tfImmutable)
+        {
+            JLOG(j.warn())
+                << "SetRemarks: Flags must be either tfImmutable or 0";
+            return temMALFORMED;
+        }
+
+        if (!remark.isFieldPresent(sfRemarkValue))
+        {
+            if (flags & tfImmutable)
+            {
+                JLOG(j.warn()) << "SetRemarks: A remark deletion cannot be "
+                                  "marked immutable.";
+                return temMALFORMED;
+            }
+            continue;
+        }
+
+        Blob const& val = remark.getFieldVL(sfRemarkValue);
+        if (val.size() == 0 || val.size() > 256)
+        {
+            JLOG(j.warn()) << "SetRemarks: RemarkValue cannot be empty or "
+                              "larger than 256 chars.";
+            return temMALFORMED;
+        }
+    }
+
+    return tesSUCCESS;
+}
+
+NotTEC
+SetRemarks::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureRemarks))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+    auto& j = ctx.j;
+
+    if (tx.isFieldPresent(sfFlags) && tx.getFieldU32(sfFlags) != 0)
+    {
+        JLOG(j.warn()) << "SetRemarks: Invalid flags set.";
+        return temMALFORMED;
+    }
```

Comment:
should only allow `lsfImmutable`?

And `temINVALID_FLAG` is better than`temMALFORMED` here